implement complex dtype for pyvista_ndarray

akaszynski opened the pull request:
### Problem

PyVista does not support assigning complex data to VTK DataSets:

```py
>>> import pyvista
>>> import numpy as np
>>> plane = pyvista.Plane(i_resolution=2, j_resolution=2)
>>> data = np.random.random((plane.n_points, 2)).view(np.complex128).ravel()
>>> plane.point_data['my_data'] = data

...

AssertionError: Complex numpy arrays cannot be converted to vtk arrays.Use real() or imag() to get a component of the array before passing it to vtk.

```

This is required for the proper implementation of #1953.

### Solution

This PR adds support for this:

```py
>>> import pyvista
>>> import numpy as np
>>> plane = pyvista.Plane(i_resolution=2, j_resolution=2)
>>> data = np.random.random((plane.n_points, 2)).view(np.complex128).ravel()
>>> plane.point_data['my_data'] = data
>>> plane.point_data['my_data']
pyvista_ndarray([0.19825972+0.52350422j, 0.17443232+0.94981268j,
                 0.55622938+0.56130788j, 0.39520853+0.09587489j,
                 0.28721739+0.04375279j, 0.59874206+0.21138555j,
                 0.25754638+0.333697j  , 0.06949282+0.16488391j,
                 0.10368154+0.77862362j])
```

Plotting complex data is implemented in the same manner as `matplotlib`: data is cast to real and the imaginary component is dropped. User is warned when this occurs (just like `matplotlib`.

adeak commented:
This seems to overlap with similar/same work you already added to https://github.com/pyvista/pyvista/pull/1953. Do you plan to revert those changes to prevent merge conflicts?

akaszynski commented:
> This seems to overlap with similar/same work you already added to #1953. Do you plan to revert those changes to prevent merge conflicts?

Yes. The main point of this PR was to separate out the features and bug fixes added in #1953 since it's scope has grown.

adeak commented on pyvista/core/datasetattributes.py:
```diff
@@ -611,6 +618,13 @@ def set_array(
+        # remove complex association if dataset is not complex
+        complex_assoc = self.dataset._association_complex_names
+        if name in complex_assoc[self.association.name]:
+            if not isinstance(data, np.ndarray) or not np.issubdtype(data.dtype, complex):
```
I know this is an unlikely edge case, but will anyone pass a list of complex values? If we want to anticipate that might be a bit safer to call `np.asarray()` on the data and check its dtype. I think `_prepare_array()` also converts iterables (should probably be sequences!) to arrays before checking complexness, so this would be safer.

And again I'd check `issubdtype(data.dtype, np.complexfloating)`, should we ever decide to allow quad precision complexes.

akaszynski replied:
I've removed this and placed the check within `_prepare_array` since we convert to a `numpy.ndarray` there already.

tkoyama010 commented on tests/plotting/test_plotting.py:
```diff
@@ -2334,3 +2334,16 @@ def test_ruler(sphere):
     plotter.add_ruler([-0.6, -0.6, 0], [0.6, -0.6, 0], font_size_factor=1.2)
     plotter.view_xy()
     plotter.show(before_close_callback=verify_cache_image)
+
+
+def test_plot_complex_value(plane):
+    """Test plotting complex data."""
+    data = np.arange(plane.n_points, dtype=np.complex128)
+    data += np.linspace(0, 1, plane.n_points) * -1j
+    with pytest.warns(np.ComplexWarning):
+        plane.plot(scalars=data)
+
+    pl = pyvista.Plotter()
+    with pytest.warns(np.ComplexWarning):
+        pl.add_mesh(plane, scalars=data, show_scalar_bar=False)
+    pl.show(before_close_callback=verify_cache_image)
```
Could you please show the scalar bar with a title to make it clear what the plot is showing?

akaszynski replied:
Added

adeak commented on pyvista/plotting/mapper.py:
```diff
@@ -103,6 +103,11 @@ def set_scalars(
 
+            # Use only the real component if an array is complex
+            if np.issubdtype(scalars.dtype, complex):
+                scalars = scalars.astype(float)
```
Don't we want to raise an informative warning of our own instead of relying on NumPy's generic warning?

akaszynski replied:
I like how it matches `matplotlib`'s behavior. Users can google it and get way more informative examples rather than having to interpret our warning.

If you disagree, I can change it.

adeak replied:
Hmm, good point, having good googlability is probably a good idea. Let's leave it like this, we can always introduce our own warning later if we get issues about the current state.

adeak commented on tests/test_datasetattributes.py:
```diff
@@ -517,3 +518,44 @@ def test_active_t_coords_name(plane):
+    ref = weakref.ref(data)
+    del data
+    assert np.allclose(plane.point_data[name], data_copy)
+    assert ref is not None
```
```suggestion
    assert ref() is not None
```

akaszynski commented on pyvista/core/dataset.py:
```diff
@@ -1420,6 +1420,7 @@ def copy_meta_from(self, ido: 'DataSet'):
         self._active_vectors_info = ido.active_vectors_info
         self.clear_textures()
         self._textures = {name: tex.copy() for name, tex in ido.textures.items()}
+        self._association_complex_names = ido._association_complex_names.copy()
```
`ido._association_complex_names` is a set, and `copy` is a "deep" copy by default and not shallow.

The rest of the meta data appears to also be a deep copy:

```py
>>> import pyvista as pv
>>> import numpy as np

>>> sphere_a = pv.Sphere()
>>> sphere_a['data'] = np.random.random((sphere_a.n_points, 3))
>>> sphere_a.active_vectors_name = 'data'

>>> sphere_b = sphere_a.copy(deep=True)
>>> sphere_b['more_data'] = np.random.random((sphere_a.n_points, 3))
>>> sphere_b.active_vectors_name = 'more_data'

>>> sphere_a.active_vectors_name
'data'
```

`sphere_a.active_vectors_name` would be `'more_data'` if it was indeed a shallow copy.

adeak replied:
I disagree on both points: `copy` is a shallow copy for native containers (although sets contain hashable and this normally immutable values, so there's no difference between a shallow and a deep copy), and `ido._association_complex_names` is not a set. It's a (default)dict of sets, and the copy only copies the dict, not the sets within.

Here's a small model of the dict-of-sets we have:
```py
>>> assoc_names = {'key': {'the value'}}  # dict of sets
... copied_assoc_names = assoc_names.copy()  # shallow copy
... copied_assoc_names['key'].add('another value')  # modify value in shallow copy
... assoc_names  # show change in original
{'key': {'another value', 'the value'}}
```
This is a shallow copy.

akaszynski replied:
Yep, you're right. PR #2779 should fix this as this is out of scope for this PR.

akaszynski commented on pyvista/core/datasetattributes.py:
```diff
@@ -547,8 +544,18 @@ def get_array(
             if type(vtk_arr) == _vtk.vtkAbstractArray:
```
Tests fail when changing to `isinstance`:
```

    @active_scalars_name.setter
    def active_scalars_name(self, name: str) -> None:
        self._raise_field_data_no_scalars_vectors()
>       dtype = self[name].dtype
E       AttributeError: 'vtkmodules.vtkCommonCore.vtkStringArray' object has no attribute 'dtype'

pyvista/core/datasetattributes.py:1112: AttributeError
```

adeak replied:
I see, thanks, because we're returning a VTK array instead of a numpy one, which breaks downstream.

So the question is then: do we really want to have this if? It seems likely that we're simply not hitting this branch in the test suite, and if one day we somehow end up with an vtkAbstractArray we'll get an error... At least in a local coverage report I have lying around, this branch was not taken.

The line was added in the huge `DataSetAttributes` PR so I couldn't figure out why it's there exactly...

adeak replied:
Of course this is also unrelated to the PR. If we want to do anything about this it will have to be another PR.

adeak replied:
Opened https://github.com/pyvista/pyvista/issues/2781 to track this.

akaszynski commented on pyvista/core/datasetattributes.py:
```diff
@@ -547,8 +544,18 @@ def get_array(
             narray = narray.view(np.bool_)  # type: ignore
+        elif name in self.dataset._association_complex_names[self.association.name]:
+            narray = narray.view(np.complex128)  # type: ignore
+            # ravel to keep this array flat to match the behavior of the rest
+            # of VTK arrays
+            if narray.ndim == 2 and narray.shape[-1] == 1:
+                narray = narray.ravel()
```
You're correct, `squeeze` accomplishes the same goal of removing singleton dimensions for 2D arrays.

akaszynski commented on pyvista/utilities/geometric_objects.py:
```diff
@@ -627,7 +627,7 @@ def Cube(center=(0.0, 0.0, 0.0), x_length=1.0, y_length=1.0, z_length=1.0, bound
 
     # add face index data for compatibility with PlatonicSolid
     # but make it inactive for backwards compatibility
-    cube.cell_data.set_array([1, 4, 0, 3, 5, 2], ['FaceIndex'])
```
Added basic checks.

adeak reviewed (approved):
One unrelated discussion (edit: marked as resolved now), and codecov claims that two lines for "data no longer complex" (lines 628 and 792) aren't hit by coverage, otherwise this LGTM, thanks!

akaszynski commented:
> codecov claims that two lines for "data no longer complex" (lines 628 and 792) aren't hit by coverage

Should be fixed now.

adeak commented on pyvista/core/dataset.py:
```diff
@@ -1420,6 +1420,7 @@ def copy_meta_from(self, ido: 'DataSet'):
         self._active_vectors_info = ido.active_vectors_info
         self.clear_textures()
         self._textures = {name: tex.copy() for name, tex in ido.textures.items()}
+        self._association_complex_names = ido._association_complex_names.copy()
```
Either https://github.com/pyvista/pyvista/pull/2779 or this should be merged first, and this extended for handling deep or shallow copies on the other PR. I recommend merging 2779 first (in time) and then handling this.

akaszynski replied:
Done in c2322ac1662ed7c18193f5dffef037203b95459a.

akaszynski replied:
Crazy that we didn't catch this earlier, but I guess people weren't using bit arrays that often. With complex it might be more of an issue.

akaszynski commented:
Merging as approved and this was a small change.

adeak commented:
> Merging as approved and this was a small change.

Please don't. My point for having no changes in without review is for a reason.

The current state is _not_ a deep copy of the dict of sets, as discussed earlier.

adeak commented on pyvista/core/dataset.py:
```diff
@@ -1426,12 +1426,16 @@ def copy_meta_from(self, ido: 'DataSet', deep: bool = True):
         self.clear_textures()
 
         if deep:
+            self._association_complex_names = ido._association_complex_names.copy()
+            self._association_bitarray_names = ido._association_bitarray_names.copy()
```
This here.

adeak replied:
I've added a test that should only pass if this is fixed.

adeak reviewed (approved):
I think I've fixed the deepcopy issue, please check https://github.com/pyvista/pyvista/commit/67c12fa2ce98595e2319b62912399ccd809e18f7. If CI is green this should be good to merge by me.

akaszynski commented:
> I think I've fixed the deepcopy issue, please check [67c12fa](https://github.com/pyvista/pyvista/commit/67c12fa2ce98595e2319b62912399ccd809e18f7). If CI is green this should be good to merge by me.

Before pulling, I had implemented:
```py
            for key, value in ido._association_complex_names.items():
                self._association_complex_names[key] = value.copy()

            for key, value in ido._association_bitarray_names.items():
                self._association_bitarray_names[key] = value.copy()
```

However, I think your approach is more concise.
